### PR TITLE
fix(bom): support digest field in structured image descriptors

### PR DESCRIPTION
**Pull request by TerryHowe**
## Problem

`imageReferenceFromMapping` rejected any structured image descriptor that carried a `digest` sibling field, with an error message saying to *"fold it into repository or tag"* — but provided no mechanism to actually do so.

Helm charts (Bitnami-style and others) commonly split the image reference across three sibling fields:

```yaml
image:
  repository: docker.io/library/postgres
  tag: "17.4"
  digest: sha256:304ab813518754228f9f792f79d6da36359b82d8ecf418096c636725f8c930ad
  pullPolicy: IfNotPresent
```

With the old behaviour the extractor errored on these descriptors, making BOM extraction and digest-pin checks impossible for any component using that pattern. It also means Dependabot tooling that scans the values files only updates `tag` and not `digest`, leaving the two fields out of sync.

## Fix

Handle `digest` analogously to how `containerSHA` is already handled for Skyhook Package CRDs: append it as `@<digest>`, producing a fully-pinned `repo:tag@digest` reference.

The `registry` field continues to error because dropping it would silently resolve to the wrong registry. Rejecting `digest` was the only case where the restriction was stricter than necessary.

## Changes

- `pkg/bom/extract.go`: remove `digest` from the error switch; collect and append it as `@<digest>` after `combineCRDTriplet`
- `pkg/bom/extract_test.go`: convert the former "digest member present" error case into two success cases covering the Bitnami-style three-field form and double-append protection

## Test plan

- [ ] `go test ./pkg/bom/... ./recipes/...` passes (verified locally)
- [ ] New test `operator image mapping with separate digest field produces tag@digest ref` exercises the three-field form end-to-end

### CODE REVIEW

File: pkg/bom/extract_test.go [hunk 1]
```diff
@@ -350,6 +350,39 @@ spec:
...
+  pinned:
+    image:
+      repository: ghcr.io/example/pinned
+      tag: v1
+      digest: sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b
+`,
+			want: []string{
+				"ghcr.io/example/pinned:v1@sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b",
+			},
+		},
```
[the remark this answers is not in the document]
Comment (TerryHowe) [reply pkg/bom/extract_test.go:424]: Done. Updated the test so `repository` carries an inline `@digest` and `digest` uses a different value; the expected result contains only the inline digest, verifying the sibling is not appended.

File: pkg/bom/extract.go [hunk 2]
```diff
@@ -276,7 +286,11 @@ func imageReferenceFromMapping(n *yaml.Node) (string, error) {
 	if name == "" {
 		return "", nil
 	}
-	return combineCRDTriplet(name, repository, tag), nil
+	ref := combineCRDTriplet(name, repository, tag)
+	if digest != "" && !strings.Contains(ref, "@") {
```
Comment (mchmarny) [inline pkg/bom/extract.go]: [P1] Validate the sibling digest before accepting the pin

A descriptor with `digest: sha256:abc` reaches this branch and becomes `repo:tag@sha256:abc`. That value false-passes the [digest gate](https://github.com/NVIDIA/aicr/blob/983c1397118d9226def92781e3ab8e4d001a485b/recipes/manifest_images_test.go#L140-L145), which checks only the `sha256:` prefix, and is then emitted as the [CycloneDX version and PURL](https://github.com/NVIDIA/aicr/blob/983c1397118d9226def92781e3ab8e4d001a485b/pkg/bom/bom.go#L353-L367). A typo therefore becomes trusted BOM data instead of failing extraction. Validate the exact `sha256:<64 lowercase hex>` shape here using the existing rule, return an invalid structured-descriptor error so BOM/mirror callers fail closed, and add malformed prefix/length/hex regression cases.

**Review by mchmarny — CHANGES_REQUESTED** [review]
One verified merge blocker is documented inline. Reviewed exact head `983c139711`; core test, lint, CLI E2E, E2E, security-scan, and BOM-freshness lanes passed. The required gate is currently red because the malware scan was cancelled after 30 minutes, and the branch is behind `main`.

File: pkg/bom/extract.go (hunk 2 above)
Comment (njhensley) [inline pkg/bom/extract.go]: **🟠 Major — Validate the digest before appending it (mirror appendContainerSHA)**

The new `digest` path appends the value raw (`ref += "@" + digest`) with no format check, unlike the sibling `containerSHA` handler `appendContainerSHA`, which validates against `containerSHARE = ^sha256:[a-f0-9]{64}$` and fails loud — whose docstring explicitly states a malformed digest "must not silently propagate into the BOM, PURL, or SBOM output." `digest: garbage` yields `repo:tag@garbage`, which passes `isLikelyImage` (any `@` satisfies `hasTagOrDigest`), lands in `ParseImageRef.Digest` verbatim, and is emitted into the `pkg:oci/…@garbage` PURL position. Since this PR targets digest-pin checks, an unvalidated digest defeats the purpose — a digest-pin gate would accept a non-digest. It also flows unescaped into `PURL()` (no percent-encoding), so a digest carrying `?`/`&`/`#`/space could corrupt or forge PURL qualifiers; validating to `sha256:<hex>` closes that too. (CodeRabbit flagged this as Major and it is still open; independently reproduced here.)

**Blast radius:** A typo/truncation or a `--set` override that lands a bogus digest in a chart's structured `image.digest` field ships a malformed `@<digest>` reference into the BOM / PURL / SBOM — the exact failure the containerSHA guard exists to prevent — and a downstream digest-pin/cosign check would treat a non-digest as a valid content address.

**Fix:** Reuse `appendContainerSHA`, which already handles the empty-digest case, the double-`@` guard, and the sha256 format check: `return appendContainerSHA(combineCRDTriplet(name, repository, tag), digest)`. Add a `digest: notasha256` test asserting the invalid-descriptor error (restoring the negative coverage lost when the old 'digest member present' test was removed).

File: pkg/bom/extract_test.go [hunk 3]
```diff
@@ -350,6 +350,40 @@ spec:
 				"ghcr.io/example/null-tag",
 			},
 		},
+		{
+			// Bitnami-style and Helm-chart patterns split the image reference
+			// into separate repository, tag, and digest sibling fields. The
+			// digest is appended as @<digest> so the extracted reference is
+			// fully pinned and digest-pin tests see the correct form.
+			name: "operator image mapping with separate digest field produces tag@digest ref",
```
Comment (njhensley) [inline pkg/bom/extract_test.go:358]: **🔵 Nitpick — Add a test for the name+repository+tag+digest (name-present) form**

Both new digest test cases omit `name`, exercising only the repository-as-name fallback. The name-present branch combined with digest append is untested.

**Blast radius:** Minor coverage gap — the interaction of the `namePresent` combineCRDTriplet path with a digest suffix has no test.

**Fix:** Add one table row, e.g. `{name: binder, repository: ghcr.io/kai, tag: v1, digest: sha256:…}` → `ghcr.io/kai/binder:v1@sha256:…`.

**Review by njhensley — COMMENTED** [review]
## 🔬 Multi-persona review

**Method:** 3 independent persona reviewers (Correctness · Supply-chain/BOM · Test-coverage/Docs) → an adversarial senior meta-reviewer re-derived every finding from the resolved code at head `983c139`. Findings below survived that pass.

**Tier legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

> **Reviewer overlap:** @coderabbitai already flagged **F1 as a Major that is still open** — this panel independently reproduced it, so treat it as a second confirmation rather than a fresh claim. **F2 and F3 are net-new.** Thanks @TerryHowe for the self-review on the test-coverage thread.

### Overall assessment

The happy path is correct, the double-`@` guard is sound, and `pkg/bom` tests pass locally. But the change mirrors the existing `containerSHA` sibling-digest handling while **dropping the validation half** — leaving digest handling both *too lax* (F1: emits a malformed non-empty digest) and *too strict* (F2: hard-errors on an empty/null digest). Both live in the same ~15-line block and are best fixed together so `digest` mirrors `tag`/`containerSHA` (tolerate empty/null when absent, validate format when present).

**Recommendation: Approve with comments** — 0 blockers, but **F1 is strongly recommended before merge** (one-line fix, real supply-chain gap).

Two findings are anchored inline (F1, F5). The three below land on unchanged lines outside the diff, so they're recorded here:

---

### 🟡 F2 — Empty/null `digest:` hard-errors the entire survey (`pkg/bom/extract.go` L250)

The empty/null tolerance (`continue`) is gated `key.Value == imageTagKey`, so only `tag: ""`/`null` is treated as the appVersion idiom. A `digest: ""` or `digest:` (null) falls through to the *"must be a non-null, non-empty scalar"* return, which propagates up and aborts the whole `ExtractImagesFromYAML` call. In the no-`registry`/full-`repository` form this PR targets, a chart shipping an unpinned `digest: ""` default would fail BOM extraction outright. (Downgraded from Major: a *full* Bitnami descriptor also carries a `registry:` sibling rejected first, so the live trigger is narrow.)

**Fix:** extend the tolerance to digest — `if (key.Value == imageTagKey || key.Value == "digest") && isNullOrEmptyScalar(value)` — plus a `digest: ""` test. Fix alongside F1.

### 🟡 F3 — Stale package-level godoc still says a digest member is rejected (`pkg/bom/extract.go` L76)

The `ExtractImagesFromYAML` godoc still reads *"…when a registry or **digest** member is present (dropping either would emit a wrong or un-pinned reference)."* The function-local godoc (~L219-223) was updated; this adjacent one wasn't, so the two now contradict. Drop "or digest" here and note the digest is folded in as `@<digest>`.

### 🔵 F6 — `IsInvalidStructuredImageDescriptor` godoc omits the registry case (`pkg/bom/extract.go` L336, pre-existing)

Enumerates "name, repository, or tag" but the same error also fires for a present `registry` member. Pre-existing drift, not introduced here — optional courtesy fix while in the file.

---

### Confirmed non-issues (examined, not flagged)

- **Double-append guard** `!strings.Contains(ref, "@")` — correct and equivalent to `appendContainerSHA`'s; the inline digest wins.
- **`name == ""` early return** correctly drops a stray digest when there's no image.
- **Removal of the old "digest member present" error test** — correct (digest is now valid); positive direction well-replaced.
- **No `context` plumbing** — bounded in-memory YAML traversal; consistent with prior repo guidance.

### Summary

| 🔴 Blocker | 🟠 Major | 🟡 Minor | 🔵 Nitpick |
|:---:|:---:|:---:|:---:|
| 0 | 1 | 2 | 2 |

**Recommendation: Approve with comments.** F1 (validate the digest — reuse `appendContainerSHA`) is the one worth doing before merge; F2/F3 ride along in the same edit region.

<sub>🤖 Assembled by a multi-persona reviewer panel + adversarial meta-review. Every finding was independently re-derived from the resolved code before inclusion.</sub>

File: pkg/bom/extract.go (hunk 2 above)
Comment (yuanchen8911) [inline pkg/bom/extract.go]: Please validate `digest` before appending it. As written, `digest: sha256:abc` becomes `repo:tag@sha256:abc`. The digest-pin gate accepts it because it checks only the `sha256:` prefix, and CycloneDX then emits the malformed digest and PURL. This is reachable through the structured image descriptor path added here and ordinary values overrides.

Require the exact `^sha256:[a-f0-9]{64}$` shape and return an invalid structured-descriptor error so BOM and mirror callers fail closed. Add regression cases for invalid prefix, length, and hex.

**Review by yuanchen8911 — CHANGES_REQUESTED** [review]
The inline digest-validation issue must be resolved before merge.

File: pkg/bom/extract.go (hunk 2 above)
Comment (TerryHowe) [reply pkg/bom/extract.go]: Done. Replaced the raw `ref += "@" + digest` with `appendContainerSHA(ref, digest)`, which validates the exact `sha256:<64 lowercase hex>` shape and returns an error for malformed values (wrong prefix, truncated hex, uppercase). Added regression tests for non-sha256 prefix, truncated hex, and uppercase hex in commit 456fb6d.

Comment (TerryHowe) [reply pkg/bom/extract.go]: Done. Used `appendContainerSHA(ref, digest)` (commit 456fb6d), which mirrors the containerSHA guard: validates `sha256:<64 lowercase hex>` and returns an error on mismatch. Added three regression cases (non-sha256 prefix, truncated hex, uppercase hex) in `TestExtractImagesFromYAML_InvalidDigestInMapping`.

File: pkg/bom/extract_test.go (hunk 3 above)
Comment (TerryHowe) [reply pkg/bom/extract_test.go:358]: Done. Added `operator image mapping with name repository tag and digest produces pinned ref` in commit 456fb6d: `{name: binder, repository: ghcr.io/kai, tag: v1, digest: sha256:...}` → `ghcr.io/kai/binder:v1@sha256:...`.

File: pkg/bom/extract.go [hunk 4]
```diff
@@ -276,7 +290,8 @@ func imageReferenceFromMapping(n *yaml.Node) (string, error) {
 	if name == "" {
 		return "", nil
 	}
-	return combineCRDTriplet(name, repository, tag), nil
+	ref := combineCRDTriplet(name, repository, tag)
+	return appendContainerSHA(ref, digest)
```
Comment (yuanchen8911) [inline pkg/bom/extract.go]: This rejects the malformed digest at the extractor boundary, but `appendContainerSHA` returns a generic `ErrCodeInvalidRequest`. Exact-head verification with `digest: sha256:abc` produces an error while `IsInvalidStructuredImageDescriptor(err)` remains `false`. Mirror discovery and BOM generation therefore take their warning-and-continue branches and can succeed with the affected image omitted.

Return an error carrying `invalidStructuredImageDescriptorError` for the `digest` field so those callers fail closed. Please also assert `IsInvalidStructuredImageDescriptor(err)` in the malformed-digest regression cases; checking only the error string does not cover this contract.

**Review by yuanchen8911 — CHANGES_REQUESTED** [review]
The digest-validation fix still does not preserve the structured-descriptor error classification required by BOM and mirror callers.

File: pkg/bom/extract.go (hunk 4 above)
Comment (TerryHowe) [reply pkg/bom/extract.go]: Done. Inlined the sha256 format check in `imageReferenceFromMapping` and wrap the failure in `invalidStructuredImageDescriptorError` (field "digest", line, column) so `IsInvalidStructuredImageDescriptor` returns true and callers fail closed. Updated all three malformed-digest regression tests to assert `IsInvalidStructuredImageDescriptor(err)` in addition to the error string (commit 54c93bc).

**Review by yuanchen8911 — APPROVED** [review]

File: pkg/bom/extract.go [hunk 5]
```diff
@@ -276,7 +292,23 @@ func imageReferenceFromMapping(n *yaml.Node) (string, error) {
 	if name == "" {
 		return "", nil
 	}
-	return combineCRDTriplet(name, repository, tag), nil
+	ref := combineCRDTriplet(name, repository, tag)
+	if digest == "" || strings.Contains(ref, "@") {
```
Comment (njhensley) [inline pkg/bom/extract.go:296]: **🔵 Nitpick — Malformed/conflicting sibling digest silently dropped when ref already inline-pinned**

The guard `if digest == "" || strings.Contains(ref, "@")` returns before the `containerSHARE` validation at L299. So when name/repository/tag already carries an inline `@digest` and a sibling `digest:` is also present, a malformed or *disagreeing* sibling is dropped with no error. Not an integrity bug: the emitted reference is always the valid inline digest (nothing malformed reaches PURL/BOM/CycloneDX), and this mirrors the accepted `appendContainerSHA` precedence. Flagging as an observability/typo-detection nit only.

**Blast radius:** Observability only — a chart-authoring typo where the two pins disagree passes silently instead of failing the survey. The emitted artifact stays correctly pinned.

**Fix:** Optional: `slog.Warn` when a non-empty sibling digest is dropped because ref already contains `@`, or move the `containerSHARE` check ahead of the `@` short-circuit to fail closed on a malformed sibling.

File: pkg/bom/extract_test.go [hunk 6]
```diff
@@ -350,6 +350,78 @@ spec:
...
```
Comment (njhensley) [inline pkg/bom/extract_test.go:358]: **🔵 Nitpick — No success test for repository+digest with tag absent (repo@digest)**

All three digest success cases carry a non-empty tag. The digest-only pin path — `combineCRDTriplet(name, "", "")` yields an untagged name, then `@digest` is folded on to produce `repo@digest` — is correct but never exercised as a success case despite 100% line coverage (it reuses the same lines as the tagged case).

**Blast radius:** A regression mis-joining the no-tag + digest form (stray `:`, dropped digest) would ship silently for charts that pin by digest only.

**Fix:** Add a table row: `repository: ghcr.io/example/pinned` + `digest: sha256:<64hex>` with no tag → want `ghcr.io/example/pinned@sha256:<64hex>`.

File: pkg/bom/extract_test.go [hunk 7]
```diff
@@ -350,6 +350,78 @@ spec:
...
+`,
+			want: []string{
+				"ghcr.io/kai/binder:v1@sha256:6c3c624b58dbbcd3c0dd82b4c53f04194d1247c6eebdaab7c610cf7d66709b3b",
+			},
+		},
+		{
+			// When the repository already carries an inline @digest, the
+			// sibling digest field must not re-append a different digest.
+			// The inline digest is preserved and the sibling is ignored.
+			name: "operator image mapping does not double-append digest when ref already has one",
```
Comment (njhensley) [inline pkg/bom/extract_test.go:413]: **🔵 Nitpick — No test for malformed sibling digest + inline @ in repository**

This double-append case uses a *well-formed* sibling digest, so the malformed-sibling-with-inline-`@` short-circuit path (see the extract.go:296 note) is unproven. Coupled to that finding.

**Blast radius:** A future refactor reordering validation ahead of the inline-`@` short-circuit would flip silent-accept to error with no test to catch the semantic change.

**Fix:** Add a case: `repository` with a valid inline `@sha256:...` plus `digest: notasha256`; assert no error and want == the inline-digest ref (pins that inline wins and the malformed sibling is ignored, not validated).

**Review by njhensley — APPROVED** [review]
## Re-review — digest support in structured image descriptors

**Method:** delta re-review (RE_REVIEW) — Correctness, Security/Supply-chain, and Test-coverage persona passes, each finding adjudicated by an adversarial senior meta-reviewer against the resolved code at `61bb270b`.
**Legend:** 🔴 Blocker · 🟠 Major · 🟡 Minor · 🔵 Nitpick

Both of my prior findings are addressed — the Major one more robustly than my original suggestion. No new blockers or majors; only three optional completeness nitpicks remain (inline).

### Prior-feedback status

| Prior finding | Disposition |
|---|---|
| 🟠 **Major** — Validate the digest before appending it | ✔️ **Addressed — better than requested** |
| 🔵 **Nitpick** — Add the name+repository+tag+digest test | ✔️ **Addressed** |

**On the Major:** the digest is now validated against `containerSHARE` (`^sha256:[a-f0-9]{64}$`) and, on mismatch, returns an `invalidStructuredImageDescriptorError` (field `"digest"`). This is stronger than my suggested `appendContainerSHA(...)` fix: `appendContainerSHA` returns a plain `ErrCodeInvalidRequest` that would **not** satisfy `IsInvalidStructuredImageDescriptor`, so `pkg/mirror/discover.go` and the BOM path would warn-and-continue instead of failing closed. Inlining the check with the structured error makes those callers fail closed. Negative coverage restored in `TestExtractImagesFromYAML_InvalidDigestInMapping` (bad prefix / truncated hex / uppercase, each asserting the classifier). Thanks for the extra rigor here — this resolves the concern @mchmarny and @yuanchen8911 raised too.

### Examined and cleared (non-issues)
- Digest regex anchored both ends (RE2 `$` = end-of-text) and value `TrimSpace`'d → no PURL-breaking chars (`?`/`&`/`#`/space/trailing junk) can reach the `pkg:oci/…@<digest>` position or the CycloneDX version field.
- No nil-deref on `digestNode` (set iff `digest != ""`).
- `combineCRDTriplet` `LastIndex(":")` handles `registry:port` hosts correctly.
- `go test ./pkg/bom/...` passes; `imageReferenceFromMapping` / `combineCRDTriplet` / `appendContainerSHA` at 100%, package 96.1%.

### Summary

```
🔴 Blocker 0  |  🟠 Major 0  |  🟡 Minor 0  |  🔵 Nitpick 3      Prior: 2/2 addressed
```

**Approving** — the remaining nitpicks are optional and non-blocking. Nice work.

**Review by mchmarny — APPROVED** [review]
